### ISSUE TITLE
Looking for contributors!

### ISSUE THREAD

**Issue by Horusiath** [y-crdt/y-crdt#217]
We're looking for contributors who will help us to provide Swift and JVM (Java/Kotlin) langage bindings for the project. This is a paid offer - you have an opportunity to work on OSS project and get paid for that! :) We have an [open collective organization](https://opencollective.com/y-collective) to handle necessary payments.

We're looking for developers who are well versed with the standards of the languages they are going to provide bindings for - naming conventions and patterns, package management, automated tests and CI/CD integration via Github Action directly on the project page. Of course having some experience with building C interop layer for these languages and managing quirks of crossing managed/unmanaged memory layer is very valuable.

The work can be realized by calling the Rust library directly via special macro libraries - this is how ywasm, [ypy](https://github.com/y-crdt/ypy) and [yrb](https://github.com/y-crdt/yrb) came to live - or by using yffi project (which exposes yrs via C FII). I've noticed existing projects like [swift-bridge](https://github.com/chinedufn/swift-bridge) and [jni](https://docs.rs/jni/latest/jni/) but their maturity for this project purposes needs to be evaluated.

/cc @dmonad

**Comment by mhmd-azeez** [comment]
I don't see a .NET port, is the offer available for .NET too? I haven't looked at properly, but I can probably write the bindings in C# and create nuget packages

**Comment by Horusiath** [comment]
@mhmd-azeez there's an unmaintained implementation of Yjs: https://github.com/yjs/ycs . However I don't know how far it's developed - for sure it's missing newest features like move operation.

The main reason for not having some official, supported .NET port is that at this point in time we don't have any calls from companies interested in sponsoring/contributing to make it happen.

**Comment by mhmd-azeez** [comment]
@Horusiath makes sense. If that changes in the future, I appreciate it if you let me know

**Comment by fivestones** [comment]
I don't have the necessary qualifications for this job, but I'm excited about Swift bindings and so I just made a small contribution at open collective :-)

**Comment by heckj** [comment]
I'd love to help do some of this - I've made an XCFramework from the Yffi C bindings (https://github.com/heckj/YrsC) as a one-off to explore and experiment. I'm more than happy to revise and contribute any of that work, and help with rolling it into any sort of CI/CD mechanisms to build releases. If you wanted to build a swift wrapper around the underlying C API, it seems like a piece that you'd need to wrangle. It'll be fine for iOS and macOS directly, but some platforms will be quirky (possibly not stable) due to Rust and it's still-nascent ability to generate libraries for some of the other platforms (Mac Catalyst, tvOS, etc).

I'd love to contribute more - I'm a touch uncertain about my ability to wrap a C API to a good Swift idiomatic API, but I'd love to give it a try. In particular, I'm not 100% sure how to handle the very dynamic nature of the library as it might apply to a more strictly typed Swift API. Anyway, interested in helping!

**Comment by nugmanoff** [comment]
Although I don't have much experience with wrapping the C APIs, I do have experience with producing software the Swifty way (all of the things you've mentioned by standards of the language, package management, incorporating CI/CD & etc).

Tbh I think with the progress he's already made @heckj is more suitable for the job, but if you need a spare hand with this one – let me know, I am looking forward to it.

P.S. As far as I can judge, [swift-bridge](https://github.com/chinedufn/swift-bridge) looks immature to me at this point of time, but we can try to play with it and see how far we get, depending on the [current state](https://github.com/chinedufn/swift-bridge#built-in-types) of interop between types. If it works – it can be leveraged as the temp solution to hit the ground running, as you did with [yrb and rutie](https://github.com/y-crdt/yrb/blob/main/docs/decisions.md#2022-05-05). 

P.P.S. But I think due to the Swift's first-class C interop support, using yffi directly would be better.

**Comment by koral--** [comment]
I can help with the JVM bindings.
I have some experience with JNI e.g. in https://github.com/koral--/android-gif-drawable and with CI/CD, mostly Bitrise: https://github.com/bitrise-io/bitrise-steplib/commits?author=koral--

**Comment by finmily** [comment]
> I can help with the JVM bindings. I have some experience with JNI e.g. in https://github.com/koral--/android-gif-drawable and with CI/CD, mostly Bitrise: https://github.com/bitrise-io/bitrise-steplib/commits?author=koral--

Hey guys. Have any plans with JVM bindings?

**Comment by Horusiath** [comment]
> Hey guys. Have any plans with JVM bindings?

There are. We already use Mozilla [uniffi](https://mozilla.github.io/uniffi-rs/) project, which can automatically generate bindings for several languages - this part is actively developed by @nugmanoff and @heckj to make first release of [y-swift](https://github.com/y-crdt/yswift). We also want to use it to support JVM via [Kotlin](https://github.com/y-crdt/ykt).

**Comment by dgg1912** [comment]
> There are. We already use Mozilla [uniffi](https://mozilla.github.io/uniffi-rs/) project, which can automatically generate bindings for several languages - this part is actively developed by @nugmanoff and @heckj for the purposes of [y-swift](https://github.com/y-crdt/yswift). We also want to use it to support JVM via [Kotlin](https://github.com/y-crdt/ykt).

@Horusiath  Looking forward to y-kotlin or y-java coming online! How is it going?

**Comment by francescotescari** [comment]
+1: I would like to work on a Kotlin multiplatform wrapper using Yjs for JS targets and yrs for android and native (ios), maybe also for wasm. I have no experience with rust though, so at least a POC of jvm bindings and uniffi setup would help me to start

**Comment by Horusiath** [comment]
@dgg1912 @francescotescari if you want, feel free to take a look at the [Kotlin](https://github.com/y-crdt/ykt) project. It's just a skeleton atm. but I can help to organise the works. We can start from working over the GitHub issues and if necessary move to our discord channel or weekly meetings.

The general approach we figured out after working with other language bindings is 2-layer architecture:

1. Use Mozilla's uniffi toolkit to generate a generic layer that wraps Rust code in core JVM/Swift primitives. This is done in [y-uniffi](https://github.com/y-crdt/yswift/blob/main/lib/src/yniffi.udl)
2. On top of that write actual code, that's idiomatic to the language of use. This is for both improving user's experience and hide underlying complexities of working with native code bindings.

I'm not sure if we are able to sponsor this work: we haven't heard from our sponsors for a while, maybe there will be new ones once we'll start working on the project.

**Comment by hcanyz** [comment]
Hey guys.  Are there any plans to provide language bindings for dart ?

**Comment by segreeeen** [comment]
https://github.com/segreeeen/Yrs4J/

I started working on Java bindings since I needed a way to use yjs datastructures in Java Spring Boot. Didn't get very far yet, but I'm having fun and will probably keep going 👍.

If someone wants to join in, please feel free to contribute.

**Comment by xiaoxustudio** [comment]
Can you come up with a binding for go otherwise I can't handle them in go

**Comment by Horusiath** [comment]
@xiaoxustudio we know about people working on go bindings but atm. I'm not sure if they are going to make their work public. I'll let you know once I'll learn more.

**Comment by faithererer** [comment]
> https://github.com/segreeeen/Yrs4J/
> 
> I started working on Java bindings since I needed a way to use yjs datastructures in Java Spring Boot. Didn't get very far yet, but I'm having fun and will probably keep going 👍.我开始开发 Java 绑定，因为我在 Java Spring Boot 中需要使用 yjs 数据结构。虽然进展不大，但我觉得很有趣，可能会继续下去 👍。
> 
> If someone wants to join in, please feel free to contribute.如果有人想参与，请随时贡献。

Thank you very much, is your repository still being maintained?

**Comment by segreeeen** [comment]
I'm still around 😄, I have less time now because we got a baby in June 2024 (time is now a very limited resource, lol), but I can help you get started if you want to contribute to the project for sure 👍. I'd live to see it being carried on!

**Comment by segreeeen** [comment]
> > https://github.com/segreeeen/Yrs4J/
> > I started working on Java bindings since I needed a way to use yjs datastructures in Java Spring Boot. Didn't get very far yet, but I'm having fun and will probably keep going 👍.我开始开发 Java 绑定，因为我在 Java Spring Boot 中需要使用 yjs 数据结构。虽然进展不大，但我觉得很有趣，可能会继续下去 👍。
> > If someone wants to join in, please feel free to contribute.如果有人想参与，请随时贡献。
> 
> Thank you very much, is your repository still being maintained?

Forgot to quote, so you get a message - sorry guys.
